Add TiCDC Debezium Protocol docs (#16703)
pingcap/docs: ticdc/ticdc-debezium.md

---
title: TiCDC Debezium Protocol
summary: Learn the concept of the TiCDC Debezium Protocol and how to use it.
---

# TiCDC Debezium Protocol

> [Debezium](https://debezium.io/) is a tool for capturing database changes. It converts each captured database change into a message called an "event" and sends these events to Kafka. Starting from v8.0.0, TiCDC supports sending TiDB changes to Kafka in Debezium format, simplifying the data link of using Debezium.

Reviewer: I am not sure what the phrase "simplifying the data link" is intended to convey here. I would suggest changing:

> Starting from v8.0.0, TiCDC supports sending TiDB changes to Kafka in Debezium format, simplifying the data link of using Debezium.

To something like:

> Starting from v8.0.0, TiCDC supports sending TiDB changes to Kafka using a Debezium style output format, simplifying migration from MySQL databases for users who had previously been using Debezium's MySQL integration.

Author: That's more clear.

## Use the Debezium message format

> When you use Kafka as the downstream sink, specify the `protocol` field of `sink-uri` as `debezium`. Then TiCDC encapsulates the Debezium message based on the event and send TiDB data change events to the downstream.

Author: ```suggestion
When you use Kafka as the downstream sink, specify the `protocol` field as `debezium` in `sink-uri` configuration. Then TiCDC encapsulates the Debezium message based on the event and send TiDB data change events to the downstream.
```

> When you use Kafka as the downstream sink, specify the `protocol` field as `debezium` in `sink-uri` configuration. Then TiCDC encapsulates the Debezium message based on the event and send TiDB data change events to the downstream.

Reviewer 2: ```suggestion
When you use Kafka as the downstream sink, specify the `protocol` field as `debezium` in `sink-uri` configuration. Then TiCDC encapsulates the Debezium message based on the event and sends TiDB data change events to the downstream.
```

> Currently, the Debezium protocol only supports Row changed events and directly ignores DDL events and WATERMARK events. A Row changed event represents a change in a row of data. When a row changes, the Row changed event is sent, including relevant information about the row both before and after the change.

Reviewer 2: ```suggestion
Currently, the Debezium protocol only supports Row changed events and directly ignores DDL events and WATERMARK events. A Row changed event represents a data change in a row. When a row changes, the Row changed event is sent, including relevant information about the row both before and after the change. A WATERMARK event marks the replication progress of a table, indicating that all events less than the watermark have been sent to the downstream.
```

Reviewer 2: Do we need to use "Row Changed Event" as [TiCDC Open Protocol | PingCAP Docs](https://docs.pingcap.com/tidb/dev/ticdc-open-protocol) does?

![image](https://github.com/pingcap/docs/assets/79440533/3651f474-214b-4ace-9ac2-e3749bb26527)

Author: Although I don't think it's necessary to capitalize all words, we can keep its usage consistent among different docs pages.

                "type":"struct",
                "optional":true,
                "name":"default.test.t2.Value",
                "field":"after",
                "fields":[
                    {
                        "type":"int32",
                        "optional":false,
                        "field":"a"
                    },
                    {
                        "type":"int32",
                        "optional":true,
                        "field":"b"
                    }
                ]
            },
            {
                "type":"string",
                "optional":false,
                "field":"op"
            },
            ...
        }
    }
}
```

The key fields of the preceding JSON data are explained as follows:

| Field      | Type   | Description                                            |
|:----------|:-------|:-------------------------------------------------------|
| payload.op        | String | The type of the change event. `"c"` indicates an `INSERT` event, `"u"` indicates an `UPDATE` event, and `"d"` indicates a `DELETE` event.  |
| payload.ts_ms     | Number | The timestamp (in milliseconds) when TiCDC generates this message. |

> | payload.before    | JSON   | The data value before the event statement change. For `"c"` events, the value of the `before` field is `null`.     |

Reviewer 2: ```suggestion
| payload.before    | JSON   | The data value before the change event of a statement. For `"c"` events, the value of the `before` field is `null`.     |
```

> | payload.after     | JSON   | The data value after the event statement change. For `"d"` events, the value of the `after` field is `null`.     |

Reviewer 2: ```suggestion
| payload.after     | JSON   | The data value after the change event of a statement. For `"d"` events, the value of the `after` field is `null`.     |
```

| payload.source.commit_ts     | Number  | The `CommitTs` identifier when TiCDC generates this message.       |
| payload.source.db     | String   | The name of the database where the event occurs.    |

> | payload.source.table     | String  |  The name of the data table where the event occurs.   |

Reviewer 2: ```suggestion
| payload.source.table     | String  |  The name of the table where the event occurs.   |
```

| schema.fields     | JSON   |  The type information of each field in the payload, including the schema information of the row data before and after the change.   |

### Data type mapping

> The data format mapping in the TiCDC Debezium message basically follows the [Debezium data type mapping rules](https://debezium.io/documentation/reference/2.4/connectors/mysql.html#mysql-data-types), which is generally consistent with the native message of the Debezium Connector. However, for some data types, the following differences exist between TiCDC Debezium and Debezium Connector messages:

Reviewer: I would suggest making it clear that we are following the Debezium type mappings used for MySQL compatibility. E.g. changing the first sentence to something like:

> The data format mapping in the TiCDC Debezium message basically follows the [Debezium data type mapping rules](https://debezium.io/documentation/reference/2.4/connectors/mysql.html#mysql-data-types), which is generally consistent with the native message of the Debezium Connector for MySQL.
